server/db/db.js

```diff
@@ -4,68 +4,114 @@
       if (err) {
+        log.warn('error saving user: ' + err);
+        // TODO send the item to a retry queue?
         return cb(err);
       }
-      conn.query(query, [buf(fxaId), email, oauthToken], function(err) {
-        if (err) {
-          log.warn('error saving user: ' + err);
-          // TODO send the item to a retry queue?
-        }
-        conn.release();
-        cb(err);
-      });
+      log.info('created user');
+      cb(err);
     });
   },
   getUserById: function(fxaId, cb) {
-    var query = 'SELECT email, oauth_token FROM users WHERE fxa_id = ?';
-    _getConn(function(err, conn) {
-      // TODO handle conn err?
-      conn.query(query, buf(fxaId), function(err, result) {
-        if (err) {
-          log.warn('error retrieving user: ' + err);
-        }
-        conn.release();
+    _trace('db.getUserById', arguments);
+    var query = 'SELECT email, oauthToken FROM users WHERE fxaId = ?';
+    pool.query(query, fxaId, function(err, result) {
+      if (err) {
+        log.warn('error retrieving user: ' + err);
+      }
 
-        cb(err, result);
-      });
+      cb(err, result);
+    });
+  },
+  getPaginatedVisits: function(fxaId, visitId, count, cb) {
+    _trace('db.getPaginatedVisits', arguments);
+    var query = 'SELECT id, url, urlHash, title, visitedAt ' +
+                'FROM visits WHERE fxaId = ? ' +
+                'and visitedAt > (SELECT visitedAt from visits where id = ?) ' +
+                'ORDER BY visitedAt LIMIT ?';
+    pool.query(query, [fxaId, visitId, count], function(err, results) {
+      cb(err, results);
+    });
+  },
+  getVisits: function(fxaId, count, cb) {
+    _trace('db.getVisits', arguments);
+    var query = 'SELECT id, url, urlHash, title, visitedAt ' +
+                'FROM visits WHERE fxaId=? ORDER BY visitedAt LIMIT ?';
+    pool.query(query, [fxaId, count], function(err, results) {
+      cb(err, results);
+    });
+  },
+  getVisit: function(fxaId, visitId, cb) {
+    _trace('db.getVisit', arguments);
+    // important: MySQL enforces that the user with id `fxaId` is the user with visit `visitId`
+    var query = 'SELECT id, url, urlHash, title, visitedAt ' +
+                'FROM visits WHERE id = ? AND fxaId = ?';
+    pool.query(query, [visitId, fxaId], function(err, r) {
+      if (err) {
+        log.warn('error getting visit: ' + err);
+      }
+      cb(err, r && r[0]);
+    });
+  },
+  // TODO if the url and visitedAt are the same, should we just discard the record?
+  // ...maybe just deal with it later
+  createVisit: function(fxaId, visitId, visitedAt, url, title, cb) {
+    _trace('db.createVisit', arguments);
+    var urlHash = crypto.createHash('sha1').update(url).digest('hex').toString();
+    var query = 'INSERT INTO visits (id, fxaId, rawUrl, url, urlHash, title, visitedAt) ' +
+                'VALUES (?, ?, ?, ?, ?, ?, ?)';
+    pool.query(query, [visitId, fxaId, url, url, urlHash, title, visitedAt], function(err, results) {
```

Comment:
Similar to my previous [stupid] comment:

```sql
CREATE TABLE IF NOT EXISTS visits (
    id CHAR(36) NOT NULL PRIMARY KEY,
    fxaId CHAR(32) NOT NULL,
    url TEXT NOT NULL,
    rawUrl TEXT NOT NULL,
    urlHash CHAR(40) NOT NULL,
    title VARCHAR(128) NOT NULL,
    visitedAt DATETIME(3) NOT NULL DEFAULT CURRENT_TIMESTAMP(3),
    updatedAt DATETIME ON UPDATE CURRENT_TIMESTAMP,
    INDEX fxaId_visitedAt_id (fxaId, visitedAt, id)
) ENGINE=InnoDB CHARACTER SET utf8 COLLATE utf8_unicode_ci;
```

Do we need to be making sure that the page `title` isn't longer than 128 chars and overflowing our field?
I think the rest of the fields are save since they're `TEXT`, dates, or computed/supplied (`fxaId`, `urlHash`).